### PR TITLE
16.0 data sources api lul

### PR DESCRIPTION
**Pull request by LucasLefevre**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/spreadsheet/static/src/commands.json [hunk 1]
```diff
@@ -333,24 +263,17 @@
...
-              ],
-              "orderBy": []
+          "definition": {
+            "id": "bb8b5846-b950-4541-8bc0-8dbe30e14b63",
+            "groupBy": [
+              "stage_id",
+              "country_id"
+            ],
+            "measure": "__count",
+            "order": "ASC",
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/commands.json:273]: I am not sure about that "order". The order is always applied on the measure.
Should it have the same shape as order clauses for pivots and lists ? even though the field is redundant.
`{ "field": "__count", "asc": true}`

File: addons/spreadsheet/static/src/commands.json [hunk 2]
```diff
@@ -333,24 +263,17 @@
...
-          "searchParams": {
-              "comparison": null,
-              "context": {},
-              "domain": [],
-              "groupBy": [
-                  "stage_id",
-                  "country_id"
-              ],
-              "orderBy": []
+          "definition": {
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/commands.json:266]: `definition.definition` ? :eyes: 
I'd rather avoid mixing standard chart related attributes from "odoo datasource" things

**Review by LucasLefevre — COMMENTED** [review]
All "datasources" share 4 attributes: id, model, domain and context

File: addons/spreadsheet/static/src/data.json [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      "context": {},
+      "domain": [],
+      "id": "1",
+      "measures": ["expected_revenue", "recurring_revenue_monthly"],
+      "model": "crm.lead",
+      "rowGroupBys": [
+        "create_date:month"
+      ],
+      "name": "Pipeline Analysis by Stage",
+      "orderBy": { "field": "expected_revenue", "asc": true},
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/data.json:83]: alternatives that exists in the code base
1. `"orderBy": { "field": "expected_revenue", "asc": true}`
2. `"orderBy": "expected_revenue ASC"`
3. `"orderBy": { "field": "expected_revenue", "order": "asc"}`

I like the simplicity and readability of 2. but there's additional logic to "parse" the string (duplicated in upgrade scripts)

Comment (LucasLefevre) [reply addons/spreadsheet/static/src/data.json:83]: 1. is the ORM api, so it's probably the way to go

File: addons/spreadsheet/static/src/commands.json [hunk 4]
```diff
@@ -243,86 +222,37 @@
...
+          "activity_ids",
+          "my_activity_date_deadline",
+          "expected_revenue",
+          "recurring_revenue_monthly",
+          "stage_id"
+        ],
+        "domain": [],
+        "model": "crm.lead",
+        "context": {},
+        "orderBy": [
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/commands.json:241]: note that orderBy for list is an array while it's a single measure for pivots and charts

File: addons/spreadsheet/static/src/data.json (hunk 3 above)
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/data.json:83]: Actually it's `{name: 'bar', asc: false}` which doesn't look good.

File: addons/spreadsheet/static/src/commands.json (hunk 2 above)
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/commands.json:266]: but I don't think of any good alternative name for the data source definition vs chart definition

Comment (pro-odoo) [reply addons/spreadsheet/static/src/commands.json:266]: `odooDefinition` ? At least it's clear that it's odoo related stuff.

File: addons/spreadsheet/static/src/commands.json (hunk 1 above)
Comment (pro-odoo) [reply addons/spreadsheet/static/src/commands.json:273]: Yes I would go for the same shape of the others

File: addons/spreadsheet/static/src/data.json (hunk 3 above)
Comment (pro-odoo) [reply addons/spreadsheet/static/src/data.json:83]: For me `2` is ok, we already have a parsing process for group bys with a date (`create_date:day`). Either we go for `2`, either we change group by to something like `{ field: "create_date", group: "day" }`.
Anyway, at runtime, it should be great to not have to parse each time we need the information

File: addons/spreadsheet/static/src/commands.json (hunk 4 above)
Comment (pro-odoo) [reply addons/spreadsheet/static/src/commands.json:241]: Is it really possible to sort with two fields in a list ? I cannot do it from the UI (or I missed something)

Comment (LucasLefevre) [reply addons/spreadsheet/static/src/commands.json:241]: Yes you can, sort several columns one after the other.
While the UI doesn't really reflect it, "previous" sorts are still taken into account and becomes "secondary sort".
Check the search_read rpc ;)

Comment (pro-odoo) [reply addons/spreadsheet/static/src/commands.json:241]: 😬

File: addons/spreadsheet/static/src/data.json (hunk 3 above)
Comment (pro-odoo) [reply addons/spreadsheet/static/src/data.json:83]: Actually we need the "groupId". It's used to sort values on a specific column. Let's take the default CRM pivot, it's different to sort by `expected_revenue` on the `total` or when `stage_id` = `Qualified`.

**Pull request closed without merging**